ecs-agent/acs/session/task_manifest_responder.go

```diff
@@ -0,0 +1,179 @@
+// whether a task manifest message is stale or not. If not, it proceeds to create a
+// list of stop candidates.
+func (tmr *taskManifestResponder) processTaskManifestMessage(
+	message *ecsacs.TaskManifestMessage) error {
+	manifestSeqNum := aws.Int64Value(message.Timeline)
+	agentLatestSeqNum := tmr.snAccessor.GetLatestSequenceNumber()
+
+	// Verify that task manifest isn't stale.
+	// The agent will keep track of the highest sequence number received in both PayloadMessage and TaskManifestMessage.
+	// If the sequence number on TaskManifestMessage is lower to the latest known one, the TaskManifestMessage must be discarded.
+	// ACS will guarantee that the sequence number is indeed increasing and valid for this use case.
+	if manifestSeqNum < agentLatestSeqNum {
+		logger.Warn(fmt.Sprintf("Discarding task manifest message. "+
+			"Sequence number from manifest is less than Agent latest sequence number."), logger.Fields{
+			"ManifestSeqNum":    manifestSeqNum,
+			"AgentLatestSeqNum": agentLatestSeqNum,
+		})
+		return nil
+	}
+
+	// Save the task manifest's messageID and sequence number so that it can be compared
+	// by the taskStopVerificationAckResponder.
+	err := tmr.messageIDAccessor.SetMessageID(aws.StringValue(message.MessageId))
+	if err != nil {
+		return errors.Wrap(err, "failed to update latest manifest messageID")
+	}
+	err = tmr.snAccessor.SetLatestSequenceNumber(message)
+	if err != nil {
+		return errors.Wrap(err, "failed to update latest manifest sequence number")
+	}
+
+	// Get list of tasks on instance.
+	tasksToMessage, err := tmr.taskComparer.CompareRunningTasksOnInstanceWithManifest(message)
+	if err != nil {
+		return err
+	}
+
+	// Acknowledge the task manifest message asynchronously.
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		tmr.ackTaskManifestMessage(message)
+	}()
+
+	// Send the task stop verification message to ACS asynchronously.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		tmr.sendTaskStopVerification(message, tasksToMessage)
+	}()
+	wg.Wait()
+	return nil
+}
+
+func (tmr *taskManifestResponder) ackTaskManifestMessage(message *ecsacs.TaskManifestMessage) {
+	messageID := aws.StringValue(message.MessageId)
+	logger.Info(fmt.Sprintf("Acking %s id: %s", TaskManifestMessageName, messageID))
+	err := tmr.respond(&ecsacs.AckRequest{
+		Cluster:           aws.String(tmr.cluster),
+		ContainerInstance: message.ContainerInstanceArn,
+		MessageId:         aws.String(messageID),
+	})
+	if err != nil {
+		logger.Error(fmt.Sprintf("Error Acking %s", TaskManifestMessageName), logger.Fields{
+			field.MessageID: messageID,
+			field.Error:     err,
+		})
+	}
+}
+
+func (tmr *taskManifestResponder) sendTaskStopVerification(message *ecsacs.TaskManifestMessage, tasksToStop []*ecsacs.TaskIdentifier) {
+	messageID := aws.StringValue(message.MessageId)
+	if len(tasksToStop) == 0 {
+		return
+	}
+	for _, task := range tasksToStop {
```

Comment:
Why do we need to send this log message for every task in the task stop verification message? Why not just log once per single task stop verification message?

Reply:
I think it is helpful here to log what tasks are being stopped for the sake of debugging so anybody debugging has visibility into what tasks are mismatched from the manifest and instance. I agree for your next comment though, we can log once if there is an error.

Reply:
> I think it is helpful here to log what tasks are being stopped for the sake of debugging so the developer has visibility into what tasks are mismatched from the manifest and instance.

That's fair - in this case could we update the log level to DEBUG then? And having a separate log statement for each task seems excessive and can potentially give off the incorrect idea that a task stop verification message is sent for each task that is stopped (e.g., if there are 5 tasks to be stopped, then someone could see 5 log statements saying "Sending task stop verification message" for the same task stop verification message, even though only one was sent in reality). Could we just log the tasks that are being stopped in a single log statement?

Reply:
hmm, on the other hand one log for 5 messages in one is messy and difficult to decipher. Could I have something like the following?
Sending task stop verification message including:
task1
task2
task3

Reply:
Update: there is no easy way to print a list in a readable way, so I have documented that we will only send a task stop verification message once despite printing possibly many times.

Reply:
> hmm, on the other hand one log for 5 messages in one is messy and difficult to decipher. 

But there aren't 5 messages? We only send one task stop verification message.

> Could I have something like the following?
Sending task stop verification message including:
task1
task2
task3

Yes, this is what I had in mind if we want to log and see what tasks are being stopped.

> there is no easy way to print a list in a readable way

What sort of difficulties are you running into? Is there anything stopping us from doing something like:

```
<get list of task ARNs>
logger.Debug("Sending task stop verification message", logger.Fields{
	field.MessageID: messageID,
	"taskARNList": <insert list of task ARNs here>,
})
```

Reply:
synced offline and this has been changed.